specification/dartLangSpec.tex

```diff
@@ -11641,6 +11765,246 @@ \subsection{Assignable Expressions}
+% An instance variable initializer is in the instance scope, so it is
+% actually possible for $D$ to be an instance member here.
+
+\commentary{%
+Note that an initializing expression for an instance variable
+is not inside such a body.%
+}
+
+\LMHash{}%
+Otherwise $\ell$ occurs in
+the body of an instance member or a generative constructor.
+In this case it is a compile-time error
+if $D$ is not an instance member and the name of $D$ is not $n$;
+and it is a compile-time error
+if $D$ is an instance member,
+and the interface of the enclosing class does not have a member named $n$.
+\EndCase
+
+\commentary{%
+Note that we are always looking up \emph{both} \id{} and \code{\id=},
+no matter whether $n$ is \id{} or \code{\id=}.%
+}
+
+\rationale{%
+This approach creates a tighter connection between a pair of declarations
+where one is a getter named \id{}
+and the other is a setter named \code{\id=}.
+This allows developers to think about
+a getter and setter that are declared together as a single entity,
+rather than two independent declarations.%
+}
+
+\commentary{%
+For example, if a term refers to \id{} and needs a setter,
+and the innermost declaration named \id{} or \code{\id=} is a getter $g$
+and there is no corresponding setter,
+it is a compile-time error.
+This error occurs even in the case where a more remote enclosing scope has
+a declaration of a setter $s$ named \code{\id=},
+because we already committed to using $g$
+(so that's actually ``a setter/getter pair where the setter is missing''),
+and we could say that this ``pair'' shadows $s$:%
+}
+
+\begin{dartCode}
+\SET{} id(int value) \{\} // \comment{This is $s$}
+\\
+\CLASS{} A \{
+  int \GET{} id => 42; // \comment{This is $g$}
+  user() \{
+    id = 0; // \comment{Compile-time error}
+  \}
+\}
+\end{dartCode}
+
+% At this point we know that
+% - $D$ does not exist, but $\ell$ has access to `this` and the
+%   interface of the enclosing class has a member named $n$, or
+% - $D$ exists and has name $n$, and $\ell$ cannot access `this`, or
+% - $D$ exists, its name can be different from $n$, but $\ell$ has access to
+%   `this`, and the interface of the enclosing class has a member named $n$.
+%
+% The lookup yields a member signature iff $\ell$ has access to `this` and
+% $D$ does not exist or $D$ is an instance member.
+%
+%
+
+\LMHash{}%
+Now proceed as described in the first applicable case from the following list:
+
+\begin{itemize}
+\item
+  When $D$ does not exist,
+  the lexical lookup yields the member signature of
+  the member of the interface of the enclosing class
+  which has the name $n$.
+  \commentary{%
+    In this case it is guaranteed that $\ell$ occurs inside
+    the body of an instance member or generative constructor,
+    and said member exists.%
+  }
+\item
+  Consider the case where $D$ is a formal type parameter declaration.
+  It is a compile-time error if $\ell$ occurs inside
+  a static method, getter, setter, or variable initializer.
```

Comment:
No longer true for a *generic* static method, which can declare its own type parameter.
So, this is only for type parameters of a class or mixin.

(Also, what about:
```dart
typedef Foo<T> = T Function(T);
```
Do we do lexical lookup on the `T`'s? If so, they also find a type parameter declaration.

Reply:
Right! Changed to say `of a class or a mixin`, which leaves type aliases and generic methods untouched.

For the `typedef` we definitely need to perform a scoped lookup:

```dart
typedef F<S> = T Function<T>(S);
```